Lib/test/test_named_expressions.py

```diff
@@ -124,12 +124,53 @@ def test_named_expression_invalid_rebinding_list_comprehension_iteration_variabl
             ("Unreachable reuse", 'i', "[False or (i:=0) for i in range(5)]"),
             ("Unreachable nested reuse", 'i',
                 "[(i, j) for i in range(5) for j in range(5) if True or (i:=10)]"),
+            # Regression tests from https://github.com/python/cpython/issues/87447
+            ("Complex expression: a", "a",
+                "[(a := 1) for a, (*b, c[d+e::f(g)], h.i) in j]"),
+            ("Complex expression: b", "b",
+                "[(b := 1) for a, (*b, c[d+e::f(g)], h.i) in j]"),
         ]
         for case, target, code in cases:
             msg = f"assignment expression cannot rebind comprehension iteration variable '{target}'"
             with self.subTest(case=case):
                 with self.assertRaisesRegex(SyntaxError, msg):
-                    exec(code, {}, {})
+                    exec(code, {}) # Module scope
+                with self.assertRaisesRegex(SyntaxError, msg):
+                    exec(code, {}, {}) # Class scope
+                with self.assertRaisesRegex(SyntaxError, msg):
+                    exec(f"lambda: {code}", {}) # Function scope
+
+    def test_named_expression_valid_rebinding_list_comprehension_iteration_variable(self):
+        cases = [
```

Comment:
Can you add a comment at the start of the test explaining that we are cheching that anything that is not directly a comprehension iteration variable (a or b) can be assigned to?

Also, we probably should ensure ALL comprehensions are covered by the test (add another for loop that changes the brackets).

Reply:
Done! Except for `dict` comprehensions. Looks like they are not tested in `test_named_expressions` at all. I've opened a new task for it here: https://github.com/python/cpython/issues/100746